File: server/world/physics/chunk_colliders.rs
use hashbrown::HashMap;
use rapier3d::prelude::{
    ColliderBuilder, ColliderHandle, ColliderSet, IslandManager,
    RigidBodySet as RapierBodySet,
};
use nalgebra::Vector3;

use crate::{Vec2, Vec3, VoxelAccess};
use super::Registry;

/// Manages static colliders for voxel chunks
/// Each solid surface voxel gets a box collider in Rapier's physics world
pub struct ChunkColliderManager {
    /// Map from chunk 2D coordinates (cx, cz) to list of collider handles
    chunk_colliders: HashMap<Vec2<i32>, Vec<ColliderHandle>>,
}

impl ChunkColliderManager {
    pub fn new() -> Self {
        Self {
            chunk_colliders: HashMap::new(),
        }
    }

    /// Generate or update colliders for a chunk region
    /// This creates box colliders for all solid, surface voxels in the specified region
    pub fn update_chunk_collider(
        &mut self,
        chunk_coords: &Vec2<i32>,
        space: &dyn VoxelAccess,
        registry: &Registry,
        collider_set: &mut ColliderSet,
        island_manager: &mut IslandManager,
        body_set: &mut RapierBodySet,
        chunk_size: usize,
        max_height: usize,
    ) {
        // Remove old colliders if they exist
        if let Some(handles) = self.chunk_colliders.remove(chunk_coords) {
            for handle in handles {
                collider_set.remove(handle, island_manager, body_set, false);
            }
        }

        let Vec2(cx, cz) = *chunk_coords;
        let mut new_handles = Vec::new();

        // Calculate world bounds for this chunk
        let min_x = cx * chunk_size as i32;
        let min_z = cz * chunk_size as i32;
        let max_x = (cx + 1) * chunk_size as i32;
        let max_z = (cz + 1) * chunk_size as i32;

        // Iterate through all voxels in the chunk
        for vx in min_x..max_x {
            for vz in min_z..max_z {
                for vy in 0..max_height as i32 {
                    let voxel_id = space.get_voxel(vx, vy, vz);
                    let block = registry.get_block_by_id(voxel_id);

                    // Skip non-solid voxels
                    if block.is_empty || block.is_fluid || block.is_passable {
                        continue;
                    }

                    // Check if this is a surface voxel (has at least one air neighbor)
                    if !Self::is_surface_voxel(space, registry, vx, vy, vz) {
                        continue;
                    }

                    // Get AABBs for this block (supports non-cube shapes)
                    let voxel_pos = Vec3(vx, vy, vz);
                    let aabbs = block.get_aabbs(&voxel_pos, space, registry);

                    // Create box colliders for each AABB
                    for aabb in aabbs {
                        let width = aabb.max_x - aabb.min_x;
                        let height = aabb.max_y - aabb.min_y;
                        let depth = aabb.max_z - aabb.min_z;

                        // Skip degenerate AABBs
                        if width <= 0.0 || height <= 0.0 || depth <= 0.0 {
                            continue;
                        }

                        // Center position of the box
                        let center_x = (aabb.min_x + aabb.max_x) / 2.0;
                        let center_y = (aabb.min_y + aabb.max_y) / 2.0;
                        let center_z = (aabb.min_z + aabb.max_z) / 2.0;

                        // Rapier uses half-extents for boxes
                        let half_extents = Vector3::new(width / 2.0, height / 2.0, depth / 2.0);

                        // Create a static box collider
                        let collider = ColliderBuilder::cuboid(
                            half_extents.x,
                            half_extents.y,
                            half_extents.z,
                        )
                        .translation(Vector3::new(center_x, center_y, center_z))
                        .build();

                        let handle = collider_set.insert(collider);
                        new_handles.push(handle);
                    }
                }
            }
        }

        // Store handles if any colliders were created
        if !new_handles.is_empty() {
            self.chunk_colliders.insert(chunk_coords.clone(), new_handles);
        }
    }

    /// Remove colliders for a chunk
    pub fn remove_chunk_collider(
        &mut self,
        chunk_coords: &Vec2<i32>,
        collider_set: &mut ColliderSet,
        island_manager: &mut IslandManager,
        body_set: &mut RapierBodySet,
    ) {
        if let Some(handles) = self.chunk_colliders.remove(chunk_coords) {
            for handle in handles {
                collider_set.remove(handle, island_manager, body_set, false);
            }
        }
    }

    /// Check if a voxel is a surface voxel (has at least one air neighbor)
    fn is_surface_voxel(
        space: &dyn VoxelAccess,
        registry: &Registry,
        vx: i32,
        vy: i32,
        vz: i32,
    ) -> bool {
        // Check all 6 neighbors
        let neighbors = [
            (vx - 1, vy, vz),
            (vx + 1, vy, vz),
            (vx, vy - 1, vz),
            (vx, vy + 1, vz),
            (vx, vy, vz - 1),
            (vx, vy, vz + 1),
        ];

        for &(nx, ny, nz) in &neighbors {
            let neighbor_id = space.get_voxel(nx, ny, nz);
            let neighbor_block = registry.get_block_by_id(neighbor_id);

            if neighbor_block.is_empty || neighbor_block.is_fluid {
                return true;
            }
        }

        false
    }

    /// Get the number of chunks with colliders
    pub fn chunk_count(&self) -> usize {
        self.chunk_colliders.len()
    }

    /// Get the total number of colliders
    pub fn collider_count(&self) -> usize {
        self.chunk_colliders.values().map(|v| v.len()).sum()
    }

    /// Clear all chunk colliders
    pub fn clear(
        &mut self,
        collider_set: &mut ColliderSet,
        island_manager: &mut IslandManager,
        body_set: &mut RapierBodySet,
    ) {
        for (_, handles) in self.chunk_colliders.drain() {
            for handle in handles {
                collider_set.remove(handle, island_manager, body_set, false);
            }
        }
    }
}

impl Default for ChunkColliderManager {
    fn default() -> Self {
        Self::new()
    }
}

File: packages/physics-engine/src/chunk-colliders.ts
import RAPIER from "@dimforge/rapier3d-compat";
import { AABB } from "@voxelize/aabb";

/**
 * Manages static colliders for voxel chunks in Rapier physics world
 * Each solid surface voxel gets a box collider in the Rapier world
 */
export class ChunkColliderManager {
  /**
   * Map from chunk coordinates (cx,cz) to list of collider handles
   */
  private chunkColliders: Map<string, number[]> = new Map();

  /**
   * Generate chunk coordinate key
   */
  private getChunkKey(cx: number, cz: number): string {
    return `${cx},${cz}`;
  }

  /**
   * Update colliders for a chunk region
   * @param chunkCoords Chunk 2D coordinates [cx, cz]
   * @param getVoxel Function to get voxel ID at world position
   * @param getBlock Function to get block properties from voxel ID
   * @param world Rapier world instance
   * @param chunkSize Size of chunk (width/depth in voxels)
   * @param maxHeight Maximum height of world
   */
  updateChunkCollider(
    chunkCoords: [number, number],
    getVoxel: (vx: number, vy: number, vz: number) => number,
    getBlock: (id: number) => {
      isEmpty: boolean;
      isFluid: boolean;
      isPassable: boolean;
      aabbs: AABB[];
    },
    world: RAPIER.World,
    chunkSize: number,
    maxHeight: number
  ): void {
    const [cx, cz] = chunkCoords;
    const key = this.getChunkKey(cx, cz);

    // Remove old colliders if they exist
    const oldHandles = this.chunkColliders.get(key);
    if (oldHandles) {
      for (const handle of oldHandles) {
        const collider = world.getCollider(handle);
        if (collider) {
          world.removeCollider(collider, false);
        }
      }
      this.chunkColliders.delete(key);
    }

    const newHandles: number[] = [];

    // Calculate world bounds for this chunk
    const minX = cx * chunkSize;
    const minZ = cz * chunkSize;
    const maxX = (cx + 1) * chunkSize;
    const maxZ = (cz + 1) * chunkSize;

    // Iterate through all voxels in the chunk
    for (let vx = minX; vx < maxX; vx++) {
      for (let vz = minZ; vz < maxZ; vz++) {
        for (let vy = 0; vy < maxHeight; vy++) {
          const voxelId = getVoxel(vx, vy, vz);
          const block = getBlock(voxelId);

          // Skip non-solid voxels
          if (block.isEmpty || block.isFluid || block.isPassable) {
            continue;
          }

          // Check if this is a surface voxel (has at least one air neighbor)
          if (!this.isSurfaceVoxel(vx, vy, vz, getVoxel, getBlock)) {
            continue;
          }

          // Get AABBs for this block (supports non-cube shapes)
          const aabbs = block.aabbs || [
            new AABB({
              minX: vx,
              minY: vy,
              minZ: vz,
              maxX: vx + 1,
              maxY: vy + 1,
              maxZ: vz + 1,
            }),
          ];

          // Create box colliders for each AABB
          for (const aabb of aabbs) {
            const width = aabb.maxX - aabb.minX;
            const height = aabb.maxY - aabb.minY;
            const depth = aabb.maxZ - aabb.minZ;

            // Skip degenerate AABBs
            if (width <= 0 || height <= 0 || depth <= 0) {
              continue;
            }

            // Center position of the box
            const centerX = (aabb.minX + aabb.maxX) / 2;
            const centerY = (aabb.minY + aabb.maxY) / 2;
            const centerZ = (aabb.minZ + aabb.maxZ) / 2;

            // Rapier uses half-extents for boxes
            const halfExtents = {
              x: width / 2,
              y: height / 2,
              z: depth / 2,
            };

            // Create a static box collider
            const colliderDesc = RAPIER.ColliderDesc.cuboid(
              halfExtents.x,
              halfExtents.y,
              halfExtents.z
            ).setTranslation(centerX, centerY, centerZ);

            const collider = world.createCollider(colliderDesc);
            newHandles.push(collider.handle);
          }
        }
      }
    }

    // Store handles if any colliders were created
    if (newHandles.length > 0) {
      this.chunkColliders.set(key, newHandles);
    }
  }

  /**
   * Remove colliders for a chunk
   */
  removeChunkCollider(
    chunkCoords: [number, number],
    world: RAPIER.World
  ): void {
    const [cx, cz] = chunkCoords;
    const key = this.getChunkKey(cx, cz);

    const handles = this.chunkColliders.get(key);
    if (handles) {
      for (const handle of handles) {
        const collider = world.getCollider(handle);
        if (collider) {
          world.removeCollider(collider, false);
        }
      }
      this.chunkColliders.delete(key);
    }
  }

  /**
   * Check if a voxel is a surface voxel (has at least one air neighbor)
   */
  private isSurfaceVoxel(
    vx: number,
    vy: number,
    vz: number,
    getVoxel: (vx: number, vy: number, vz: number) => number,
    getBlock: (id: number) => { isEmpty: boolean; isFluid: boolean }
  ): boolean {
    // Check all 6 neighbors
    const neighbors = [
      [vx - 1, vy, vz],
      [vx + 1, vy, vz],
      [vx, vy - 1, vz],
      [vx, vy + 1, vz],
      [vx, vy, vz - 1],
      [vx, vy, vz + 1],
    ];

    for (const [nx, ny, nz] of neighbors) {
      const neighborId = getVoxel(nx, ny, nz);
      const neighborBlock = getBlock(neighborId);

      if (neighborBlock.isEmpty || neighborBlock.isFluid) {
        return true;
      }
    }

    return false;
  }

  /**
   * Get the number of chunks with colliders
   */
  get chunkCount(): number {
    return this.chunkColliders.size;
  }

  /**
   * Get the total number of colliders
   */
  get colliderCount(): number {
    let count = 0;
    for (const handles of this.chunkColliders.values()) {
      count += handles.length;
    }
    return count;
  }

  /**
   * Clear all chunk colliders
   */
  clear(world: RAPIER.World): void {
    for (const handles of this.chunkColliders.values()) {
      for (const handle of handles) {
        const collider = world.getCollider(handle);
        if (collider) {
          world.removeCollider(collider, false);
        }
      }
    }
    this.chunkColliders.clear();
  }
}
